src/translator/cache.h

```diff
@@ -0,0 +1,171 @@
+
+  CacheStats stats() const;
+
+ private:
+  void debug(std::string) const;
+
+  /// cacheConfig {sizeInBytes, recordTimeToLive, removeExpired} determins the upper limit on cache storage, time for a
+  /// record to live and whether or not to evict expired records.
+  L4::HashTableConfig::Cache cacheConfig_;
+
+  /// epochManagerConfig_{epochQueueSize, epochProcessingInterval, numActionQueues}:
+  /// - epochQueueSize (How many actions to keep in a Queue)
+  /// - epochProcessingInterval (Reclaimation runs in a background thread. The interval at which this thread is run).
+  /// - numActionQueues (Possibly to increase throughput of registering actions by a finer-granularity of locking).
+  /// .. and "action" is an std::function<void(void)>..., which is run by the background thread.
+  L4::EpochManagerConfig epochManagerConfig_;
+
+  /// L4 Service. There's a string key based registration through service_ and accesses from multiple processes to get
+  /// the same map.
+  L4::LocalMemory::HashTableService service_;
+
+  /// An L4 Context, does magic with epochs everytime accessed;  "Once a context is retrieved, the operations such as
+  /// operator[] on the context and Get() are lock-free."
+  L4::LocalMemory::Context context_;
+
+  /// context_[hashTableIndex_] gives the hashmap for Get(...) or Add(...) operations
+  size_t hashTableIndex_;
+};
+
+#endif
+
+/// Alternative cache for non-thread based workflow (specifically WASM). LRU Eviction Policy. Uses a lot of std::list.
+/// Yes, std::list; If someone needs a more efficient version, look into threads and go for ThreadSafeL4Cache.
+/// This class comes as an afterthought, so expect parameters which may not be used to achieve parity with
+/// ThreadSafeL4Cache.
+
+class ThreadUnsafeLRUCache {
+ public:
+  /// Construct a ThreadUnSafeLRUCache from configuration passed via key values in an Options object:
+  ///
+  /// @param [in] options: Options object which is used to configure the cache. See command-line parser for
+  /// documentation (`marian::bergamot::createConfigParser()`)
+  ThreadUnsafeLRUCache(Ptr<Options> options);
+
+  /// Fetches a record from cache, storing it in processedRequestSentence if found. Not thread-safe.
+  ///
+  /// @param [in]  words: query marian:Words for which translation results are looked up in cache.
+  /// @param [out] processedRequestSentence: stores cached results for use outside if found.
+  ///
+  /// @returns true if query found in cache false otherwise.
+  bool fetch(const marian::Words &words, ProcessedRequestSentence &processedRequestSentence);
+
+  /// Inserts a new record into cache.
+  ///
+  /// @param [in] words: marian::Words processed from a sentence
+  /// @param [in] processedRequestSentence: minimum translated information corresponding to words
+  void insert(const marian::Words &words, const ProcessedRequestSentence &processedRequestSentence);
+
+  CacheStats stats() const;
+
+ private:
+  struct Record {
+    marian::Words key;
```

Comment:
There's no comparison because of hash_fn, assuming we "probably" can and want to optimize for storage, will replace these with `uint64_t`. 

I will throw this (LRU) cache out for a hazard-pointer-based implementation eventually when I can prioritize this.

Reply:
Replaced with uint64_t